doc/man/popauth.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH POPAUTH 1 "November 20th, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
popauth \- Setup APOP passwords 
.SH SYNOPSIS
.B popauth
.RI [ options ]
.SH DESCRIPTION
This manual page documents briefly the
.B popauth
command.
Default action is to convert plaintext to DBM file.

This man page is not intended to be complete.
The programs are documented fully by
.IR "mailutils" ,
available via the Info system.

.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-f, \-\-file=FILE
Read input from FILE (default stdin)
.TP
.B \-o, \-\-output=FILE
Direct output to FILE
.TP
.B \-l, \-\-list
List the contents of DBM file
.TP
.B \-h, \-\-help
Display this help and exit
.TP
.B \-v, \-\-version
Display program version
.br
.SH AUTHOR
This manual page was written by Jeff Bailey <jbailey@gnu.org>.
Copyright (C) 2001 Free Software Foundation.  This document is
in the PUBLIC DOMAIN.


NEWS
GNU mailutils NEWS -- history of user-visible changes.  2003-02-12
Copyright (C) 2002, 2003 Free Software Foundation, Inc.
See the end for copying conditions.

Please send mailutils bug reports to <bug-mailutils@gnu.org>.

Version 0.3:

* General:

** Added NLS support. Currently Catalan and Spanish translations
are available.

** Fixed compilation with the newer versions of Berkeley DB.

* libmailbox:

** Functions for generating References: and In-Reply-To: headers as
per RFC 2822.

** A set of functions for sequential access to streams.

** The function mu_errstring is renamed to mu_strerror for consistency
with the usual practice.

** Added initial implementation of message header encoding and decoding
functions (RFC 2047).

* libsieve:

** The syntax of multiline strings is extended. It is possible
to use arbitrary ending delimiter instead of the default dot,
and to strip off the leading tabs, which allows for natural
indentation of the multiline strings.

** Added support for relational tests according to RFC 3431.

** Fixed incorrect handling of multiple addresses in `address'
tests.

* imap4d:

** The untagged output was fixed to conform to the RFC 2060. Specifically,
the output of STORE command is corrected in accordance with the
rfc2060-errata document; the EXIST and RECENT untagged are always
returned by EXAMINE, SELECT, and NOOP.

** Added support for TLS and GSASL.

** Implemented LOGINDISABLED capability.

** Implemented IDLE command (RFC 2177).

* pop3d: Added TLS support.

* mail: Fixed error diagnostics and removed startup banner.

* mh: Initial implementation. 


Version 0.2:

* libmailbox: Several bugfixes. The configuration suite allows
  to disable support for any protocol or mailbox format to reduce
  the size of the library.

* Added new utility mailutils-config. This utility prints gcc command
  line options needed for compiling and linknig an application against
  mailutils.

* libsieve: New library. Supports Sieve language as described in
  RFC 3028 and provides a mechanism for dynamic loading of
  user-defined actions, tests and comparators.

* sieve: Completely rewritten using libsieve.

* sieve.scm: Added vacation extension action.

* mail.local: Allows to apply user-defined sieve filters before delivering
  messages.

* mail utility: Added `editheaders' boolean variable. When set, the
  user is allowed to edit headers of the message.
  Message set specification allows negation of search primitives, e.g.
  `from ! subject:/daily/' outputs all messages whose subject
  does not contain the word 'daily'.
  Reply command correctly adds `In-Reply-To' and `References' headers
  to outgoing messages.
  When used with a numeric argument '=' command sets the current
  message to this number.
  New command `setenv' allows to set and examine environment
  variables.
  The semantics of `crt' variable is made compatible with that of
  other implementations.
  Fixed several bugs.

* libmu_scm: Added support for Guile 1.7.0


Version 0.1:

The first release.


----------------------------------------------------------------------
Copyright information:

Copyright (C) 2002 Free Software Foundation, Inc.

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.

Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


pop3d/popauth.c
/* GNU mailutils - a suite of utilities for electronic mail
   Copyright (C) 1999, 2000, 2001 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

#include "pop3d.h"

int db_list (char *input_name, char *output_name);
int db_make (char *input_name, char *output_name);

#define ACT_CREATE  0
#define ACT_ADD     1
#define ACT_DELETE  2
#define ACT_LIST    3
#define ACT_CHPASS  4

struct action_data {
  int action;
  char *input_name;
  char *output_name;
  char *username;
  char *passwd;
};

void check_action(int action);
int action_create __P((struct action_data *ap));
int action_add __P((struct action_data *ap));
int action_delete __P((struct action_data *ap));
int action_list __P((struct action_data *ap));
int action_chpass __P((struct action_data *ap));

int (*ftab[]) __P((struct action_data *)) = {
  action_create,
  action_add,
  action_delete,
  action_list,
  action_chpass
};

const char *argp_program_version = "popauth (" PACKAGE_STRING ")";
static char doc[] = "GNU popauth -- manage pop3 authentcation database";
static error_t popauth_parse_opt  __P((int key, char *arg,
				       struct argp_state *astate));

void popauth_version __P((FILE *stream, struct argp_state *state));
void (*argp_program_version_hook) __P((FILE *stream,
				       struct argp_state *state)) =
                                   popauth_version;

static struct argp_option options[] = 
{
  { NULL, 0, NULL, 0, "Actions are:", 1 },
  { "add", 'a', 0, 0, "Add user", 1 },
  { "modify", 'm', 0, 0, "Modify user's record (change password)", 1 },
  { "delete", 'd', 0, 0, "Delete user's record", 1 },
  { "list", 'l', 0, 0, "List the contents of DBM file", 1 },

  { NULL, 0, NULL, 0,
    "Default action is:\n"
    "  For the file owner: --list\n"
    "  For a user: --modify --username <username>\n", 2 },
  
  { NULL, 0, NULL, 0, "Options are:", 3 },
  { "file", 'f', "FILE", 0, "Read input from FILE (default stdin)", 3 },
  { "output", 'o', "FILE", 0, "Direct output to file", 3 },
  { "password", 'p', "STRING", 0, "Specify user's password", 3 },
  { "user", 'u', "USERNAME", 0, "Specify user name", 3 },
  { NULL, }
};

static struct argp argp = {
  options,
  popauth_parse_opt,
  NULL,
  doc,
  NULL,
  NULL, NULL
};

static const char *popauth_argp_capa[] = {
  "common",
  "license",
  NULL
};

static error_t
popauth_parse_opt (int key, char *arg, struct argp_state *astate)
{
  struct action_data *ap = astate->input;
  switch (key)
    {
    case ARGP_KEY_INIT:
      memset (ap, 0, sizeof(*ap));
      ap->action = -1;
      break;
      
    case 'a':
      check_action (ap->action);
      ap->action = ACT_ADD;
      break;
	
    case 'l':
      check_action (ap->action);
      ap->action = ACT_LIST;
      break;
	
    case 'd':
      check_action (ap->action);
      ap->action = ACT_DELETE;
      break;
	  
    case 'p':
      ap->passwd = optarg;
      break;
      
    case 'm':
      check_action (ap->action);
      ap->action = ACT_CHPASS;
      break;
	
    case 'f':
      ap->input_name = optarg;
      break;
	  
    case 'o':
      ap->output_name = optarg;
      break;
	
    case 'u':
      ap->username = optarg;
      break;
	
    case ARGP_KEY_FINI:
      if (ap->action == -1)
	{
	  /* Deduce the default action */
	  if (getuid () == 0)
	    ap->action = ACT_LIST;
	  else
	    ap->action = ACT_CHPASS;
	}
      
    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

int
main(int argc, char **argv)
{
  struct action_data adata;
  
  mu_argp_parse (&argp, &argc, &argv, 0,
		 popauth_argp_capa, NULL, &adata);

  return (*ftab[adata.action]) (&adata);
}

void
check_action (int action)
{
  if (action != -1)
    {
      mu_error ("You may not specify more than one `-aldp' option");
      exit (1);
    }
}

int
check_user_perm (int action, struct action_data *ap)
{
  struct stat sb;
  struct passwd *pw;
  uid_t uid;
  
  if (!ap->input_name)
    ap->input_name = APOP_PASSFILE;

  if (mu_dbm_stat (ap->input_name, &sb))
    {
      if (ap->action == ACT_ADD)
	{
	  DBM_FILE db;
	  if (mu_dbm_open (ap->input_name, &db, MU_STREAM_CREAT, 0600))
	    {
	      mu_error ("can't create %s: %s",
			ap->input_name, strerror (errno));
	      exit (1);
	    }
	  mu_dbm_close (db);
	  mu_dbm_stat (ap->input_name, &sb);
	}
      else
	{
	  mu_error ("can't stat %s: %s", ap->input_name, strerror (errno));
	  exit (1);
	}
    }

  uid = getuid ();
  if (uid == 0 || sb.st_uid == uid)
    return 0;

  if (ap->username)
    {
      mu_error ("Only the file owner can use --username");
      exit (1);
    }

  if (action != ACT_CHPASS)
    {
      mu_error ("Operation not allowed");
      exit (1);
    }
  pw = getpwuid (uid);
  if (!pw)
    exit (1);
  ap->username = pw->pw_name;
  return 1;
}

int
action_list (struct action_data *ap)
{
  FILE *fp;
  DBM_FILE db;
  DBM_DATUM key;
  DBM_DATUM contents;
  
  check_user_perm (ACT_LIST, ap);
  if (mu_dbm_open (ap->input_name, &db, MU_STREAM_READ, 0600))
    {
      mu_error("can't open %s: %s", ap->input_name, strerror (errno));
      return 1;
    }
  
  if (ap->output_name)
    {
      fp = fopen (ap->output_name, "w");
      if (!fp)
	{
	  mu_error("can't create %s: %s", ap->output_name, strerror (errno));
	  return 1;
	}
    }
  else
    fp = stdout;

  if (ap->username)
    {
      memset (&key, 0, sizeof key);
      memset (&contents, 0, sizeof contents);
      MU_DATUM_PTR (key) = ap->username;
      MU_DATUM_SIZE (key) = strlen (ap->username);
      if (mu_dbm_fetch (db, key, &contents))
	{
	  mu_error ("no such user: %s", ap->username);
	}
      else
	fprintf (fp, "%.*s: %.*s\n",
		 (int) MU_DATUM_SIZE (key),
		 (char*) MU_DATUM_PTR (key),
		 (int) MU_DATUM_SIZE (contents),
		 (char*) MU_DATUM_PTR (contents));
    }
  else
    {
      for (key = mu_dbm_firstkey (db); MU_DATUM_PTR(key);
	   key = mu_dbm_nextkey (db, key))
	{
	  memset (&contents, 0, sizeof contents);
	  mu_dbm_fetch (db, key, &contents);
	  fprintf (fp, "%.*s: %.*s\n",
		   (int) MU_DATUM_SIZE (key),
		   (char*) MU_DATUM_PTR (key),
		   (int) MU_DATUM_SIZE (contents),
		   (char*) MU_DATUM_PTR (contents));
	}
    }
  
  mu_dbm_close (db);
  fclose (fp);
  return 0;
}

int
action_create (struct action_data *ap)
{
  FILE *fp;
  DBM_FILE db;
  DBM_DATUM key;
  DBM_DATUM contents;
  char buf[256];
  int line = 0;

  /* Make sure we have proper privileges if popauth is setuid */
  setuid (getuid ());
  
  if (ap->input_name)
    {
      fp = fopen (ap->input_name, "r");
      if (!fp)
	{
	  mu_error("can't open %s: %s", ap->input_name, strerror (errno));
	  return 1;
	}
    }
  else
    {
      ap->input_name = "";
      fp = stdin;
    }
  
  if (!ap->output_name)
    ap->output_name = APOP_PASSFILE;
  if (mu_dbm_open (ap->output_name, &db, MU_STREAM_CREAT, 0600))
    {
      mu_error("can't create %s: %s", ap->output_name, strerror (errno));
      return 1;
    }

  line = 0;
  while (fgets (buf, sizeof buf - 1, fp))
    {
      int len;
      int argc;
      char **argv;

      len = strlen (buf);
      if (buf[len-1] == '\n')
	buf[--len] = 0;
      
      line++;
      if (argcv_get (buf, ":", NULL, &argc, &argv))
	{
	  argcv_free (argc, argv);
	  continue;
	}

      if (argc == 0 || argv[0][0] == '#')
	{
	  argcv_free (argc, argv);
	  continue;
	}
      
      if (argc != 3 || argv[1][0] != ':' || argv[1][1] != 0)
	{
	  mu_error ("%s:%d: malformed line", ap->input_name, line);
	  argcv_free (argc, argv);
	  continue;
	}

      memset (&key, 0, sizeof key);
      memset (&contents, 0, sizeof contents);
      MU_DATUM_PTR (key) = argv[0];
      MU_DATUM_SIZE (key) = strlen (argv[0]);
      MU_DATUM_PTR (contents) = argv[2];
      MU_DATUM_SIZE (contents) = strlen (argv[2]);

      if (mu_dbm_insert (db, key, contents, 1))
	mu_error ("%s:%d: can't store datum", ap->input_name, line);

      argcv_free (argc, argv);
    }
  mu_dbm_close (db);
  fclose (fp);
  return 0;
}

int
open_io (int action, struct action_data *ap, DBM_FILE *db, int *not_owner)
{
  int rc = check_user_perm (action, ap);
  if (not_owner)
    *not_owner = rc;
  if (mu_dbm_open (ap->input_name, db, MU_STREAM_RDWR, 0600))
    {
      mu_error("can't open %s: %s", ap->input_name, strerror (errno));
      return 1;
    }
  return 0;
}

void
fill_pass (struct action_data *ap)
{
  if (!ap->passwd)
    {
      char *p;

      while (1) {
	if (ap->passwd)
	  free (ap->passwd);
	p = getpass ("Password:");
	if (!p)
	  exit (1);
	ap->passwd = strdup (p);
	p = getpass ("Confirm :");
	if (strcmp (ap->passwd, p) == 0)
	  break;
	mu_error ("Passwords differ. Please retry.");
      } 
    }
}

int
action_add (struct action_data *ap)
{
  DBM_FILE db;
  DBM_DATUM key;
  DBM_DATUM contents;
  int rc;
  
  if (!ap->username)
    {
      mu_error ("missing username to add");
      return 1;
    }

  if (open_io (ACT_ADD, ap, &db, NULL))
    return 1;

  fill_pass (ap);
  
  memset (&key, 0, sizeof key);
  memset (&contents, 0, sizeof contents);
  MU_DATUM_PTR (key) = ap->username;
  MU_DATUM_SIZE (key) = strlen (ap->username);
  MU_DATUM_PTR (contents) = ap->passwd;
  MU_DATUM_SIZE (contents) = strlen (ap->passwd);

  rc = mu_dbm_insert (db, key, contents, 1);
  if (rc)
    mu_error ("can't store datum");

  mu_dbm_close (db);
  return rc;
}

int
action_delete (struct action_data *ap)
{
  DBM_FILE db;
  DBM_DATUM key;
  int rc;
  
  if (!ap->username)
    {
      mu_error ("missing username to delete");
      return 1;
    }

  if (open_io (ACT_DELETE, ap, &db, NULL))
    return 1;
  
  MU_DATUM_PTR (key) = ap->username;
  MU_DATUM_SIZE (key) = strlen (ap->username);

  rc = mu_dbm_delete (db, key);
  if (rc)
    mu_error ("can't remove record for %s", ap->username);

  mu_dbm_close (db);
  return rc;
}

int
action_chpass (struct action_data *ap)
{
  DBM_FILE db;
  DBM_DATUM key;
  DBM_DATUM contents;
  int rc;
  int not_owner;
  
  if (open_io (ACT_CHPASS, ap, &db, &not_owner))
    return 1;

  if (!ap->username)
    {
      mu_error ("missing username");
      return 1;
    }

  memset (&key, 0, sizeof key);
  memset (&contents, 0, sizeof contents);

  MU_DATUM_PTR (key) = ap->username;
  MU_DATUM_SIZE (key) = strlen (ap->username);
  if (mu_dbm_fetch (db, key, &contents))
    {
      mu_error ("no such user: %s", ap->username);
      return 1;
    }

  if (not_owner)
    {
      char *oldpass, *p;
      
      oldpass = xmalloc (MU_DATUM_SIZE (contents) + 1);
      memcpy (oldpass, MU_DATUM_PTR (contents), MU_DATUM_SIZE (contents));
      oldpass[MU_DATUM_SIZE (contents)] = 0;
      p = getpass ("Old Password:");
      if (!p)
	return 1;
      if (strcmp (oldpass, p))
	{
	  mu_error ("Sorry");
	  return 1;
	}
    }

  fill_pass (ap);
  
  MU_DATUM_PTR (contents) = ap->passwd;
  MU_DATUM_SIZE (contents) = strlen (ap->passwd);
  rc = mu_dbm_insert (db, key, contents, 1);
  if (rc)
    mu_error ("can't replace datum");

  mu_dbm_close (db);
  return rc;
}

void
popauth_version (FILE *stream, struct argp_state *state)
{
#if defined(WITH_GDBM)
# define FORMAT "GDBM"
#elif defined(WITH_BDB2)
# define FORMAT "Berkeley DB"  
#elif defined(WITH_NDBM)
# define FORMAT "NDBM"  
#elif defined(WITH_OLD_DBM)
# define FORMAT "Old DBM"  
#endif
  printf ("%s\n", argp_program_version);
  printf ("Database format: %s\n", FORMAT);
  printf ("Database location: %s\n", APOP_PASSFILE);
  exit (EXIT_SUCCESS);
}
